chain/chain/src/chain.rs

```diff
@@ -2887,32 +2890,102 @@ impl<'a> ChainUpdate<'a> {
                             shard_id,
                         )?;
 
-                    // Apply transactions and receipts.
-                    let apply_result = self
-                        .runtime_adapter
-                        .apply_transactions(
-                            shard_id,
-                            chunk_inner.prev_state_root(),
-                            chunk_header.height_included(),
-                            block.header().raw_timestamp(),
-                            &chunk_header.prev_block_hash(),
-                            &block.hash(),
-                            &receipts,
-                            chunk.transactions(),
-                            chunk_inner.validator_proposals(),
-                            prev_block.header().gas_price(),
-                            gas_limit,
-                            &block.header().challenges_result(),
-                            *block.header().random_value(),
-                            true,
-                            is_first_block_with_chunk_of_version,
-                            #[cfg(feature = "sandbox")]
-                            self.states_to_patch.take(),
-                            #[cfg(not(feature = "sandbox"))]
-                            None,
-                        )
-                        .map_err(|e| ErrorKind::Other(e.to_string()))?;
+                    let runtime_adapter = self.runtime_adapter.clone();
+                    let block_hash = block.hash().clone();
+                    let challenges_result = block.header().challenges_result().clone();
+                    let block_timestamp = block.header().raw_timestamp();
+                    let gas_price = prev_block.header().gas_price();
+                    let random_seed = *block.header().random_value();
+                    let height = chunk_header.height_included();
+                    let prev_block_hash = chunk_header.prev_block_hash().clone();
```

Comment:
This feels somewhat ugly, but I also don't see a better way to do it :(

Reply:
Tbh this is not ready for review, I just wanted to run all tests

Reply:
This probably isnt' applicable in this case, but one cool thing that rayon does is it allows to close over stack data without cloning it:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f9af6bb5b8554343b310d8012a293edf